ERCS/erc-7272.md

```diff
@@ -0,0 +1,221 @@
+- The authorization has been given by a valid issuer, i.e the EAT has been signed by one of the expected issuers
+
+
+### Structure of an Ethereum Access Token
+
+An Ethereum Access Token is composed of a signature and expiry.
+
+```
+{
+ uint8 v,
+ bytes32 r,
+ bytes32 s,
+ uint256 expiry
+}
+```
+
+The signature is obtained using the typed structured data hashing and signing standard (EIP-712), signing over the following EAT payload:
+
+```
+struct AccessToken {
+    uint256 expiry;
+    FunctionCall functionCall;
+}
+
+struct FunctionCall {
+    bytes4 functionSignature;
+    address target;
+    address caller;
+    bytes parameters;
+}
+```
+
+- **expiry**: unix timestamp, expected to be before `block.timestamp`
+
+`FunctionCall` parameters correspond to the following:
+
+- **functionSignature**: identifier for the function being called, expected to match `msg.sig`
+- **target**: address of the target contract being called
+- **caller**: address of the current caller - expected to match `msg.sender`
+- **parameters**: `calldata` after stripping off the first parameters, namely `v`,`r`, `s` and `expiry`
+
+
+### EAT Verification
+
+On chain, two contracts are necessary: an `AccessTokenConsumer` which is inherited by contracts needing to permission some of its functions and an `AccessTokenVerifier` which is responsible for verifying EATs.
+
+#### Using AccessTokenConsumer
+A modifier is added to functions that need to be permissioned with an EAT.
+For example, given a function such as:
+
+```
+function transfer(address recipient, uint256 amount)
+```
+
+Its permissioned version, protected via an EAT becomes:
+
+```
+function transfer(
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        uint256 expiry,
+        address recipient,
+        uint256 amount
+        )
+        requiresAuth(v, r, s, expiry) {
+            ...
+        }
+
+```
+
+The parameters `v`, `r`, `s`, `expiry`, which form an EAT, MUST be placed first, before any other parameters.
+
+
+The `requiresAuth` modifier is inherited from `AccessTokenConsumer` which verifies the Ethereum Access Token and consumes it:
+
+```
+    modifier requiresAuth(
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        uint256 expiry
+    ) {
+        // VF -> Verification Failure
+        require(verify(v, r, s, expiry), "AccessToken: VF");
+        _consumeAccessToken(v, r, s, expiry);
+        _;
+    }
+```
+
+The `verify` function calls the `AccessTokenVerifier` registered upon construction to verify the EAT's integrity.
+
+Optionally, implementers can decide to make EATs non-replayable. This what `_consumeAccessToken(...)` does. It marks an EAT as consumed to prevent its re-use:
+
+```
+function _consumeAccessToken(
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        uint256 expiry
+    ) private {
+        bytes32 accessTokenHash = keccak256(abi.encodePacked(v, r, s, expiry));
+
+        _accessTokenUsed[accessTokenHash] = true;
+    }
+```
```

Comment:
A lot of this is implementation details. Only the externally visible behaviour of the contract should be specified. Normally this amounts to the public functions or interface of a contract.

In this particular EIP, the order of function parameters and the structure of the EIP-712 signed message are definitely externally visible, but I don't see why an implementer would _need_ to, for example, use a modifier or inherit from `AccessTokenConsumer`.

Reply:
Point well taken.

Reply:
Pushed an update leaving out implementation details: https://github.com/ethereum/ERCs/pull/92/commits/3a37dddab647d166054e225cae501e429aa443ff